### ISSUE TITLE
Track disable/enable behavior questions

### ISSUE THREAD

**Issue by kwindla** [twilio/twilio-video.js#354]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

Hi. We've been using twilio-video since the early releases. I'm upgrading to 1.11.1 and refactoring some of our code, and have some questions about expected behavior for disabled tracks. Everything might very well be behaving as intended, but we've worked around some of this in earlier versions of our code, and now I'm trying to actually take the time to understand what's going on and write clean code! Thanks for all the great work on the libraries and product.

### (1) Calling disable() makes a video track unplayable

Calling disable() on a LocalVideoTrack sets muted to true for the receiving side's native MediaStreamTrack, and -- in Chrome at least -- if you use this muted track as part of a stream handed to src or srcObject of an HTML video element, a call to play() will not complete while the track is in this state.

This is different from how a native MediaStreamTrack added to a Chrome-to-Chrome PeerConnection works. In the native case, setting enabled to false on the sending side does not change the muted property on the receiving side, and does not change whether streams containing the track are `play()`able.

On the other hand, calling disable() on a LocalAudioTrack does not seem to set muted to true for the MediaStreamTrack on the receiving side, or to effect `play()`.

This LocalVideoTrack.disable() behavior causes problems for us because we pair audio and video tracks together into streams as early as we have enough information to do so, and then construct video elements to play those streams whenever our display layouts and compositing states change. Our assumptions from experience with peer-to-peer (Chrome-to-Chrome) streams are leading us astray when we port our code to pipe streams through twilio-video Rooms.

**Expected behavior:**

Calling disable() on a LocalVideoTrack would result in the track being playable on the remote side, but rendering as all black. The muted field of the MediaStreamTrack on the receiving side would be unchanged (value would be `false`).

**Actual behavior:**

Calling disable() on a LocalVideoTrack results in any stream containing the track not being playable on the remote side. The promise returned by `play()` neither resolves or rejects. The muted field of the MediaStreamTrack on the receiving side is set to true. A video that is already playing a stream containing this track is able to continue playing (audio tracks remain audible). But new calls to `play()` do not complete until `enable()` is called on the sending side.

### (2) Local\*Track enable()/disable() behavior depend on the enabled state of the underlying MediaStreamTrack

Calling enable() or disable() on a LocalVideoTrack or LocalAudioTrack that has an underlying MediaStreamTrack that already has its enabled field set to true or false (as the case may be) seems not to trigger enabled and disabled events on the remote side.

We use tracks in several contexts simultaneously (for example, local video/audio display and compositing elements). So there are parts of our code that call enable/disable directly on MediaStreamTrack objects that the twilio-video library also holds a reference to.

We are finding that we have to be careful about the order in which we set the enabled field and call enable()/disable(). In some cases, our code is structured so that doing this properly in all instances is not possible. :-) In the past, we have experimented with cloning tracks but found that Chrome seemed to be less stable when we relied heavily on cloned tracks/streams. Those tests may now be out of date, though.

**Expected behavior:**

My expectation was that calling enable() or disable() on a LocalAudioTrack or LocalVideoTrack would cause the remote enabled/disabled events to fire no matter the state of the underlying MediaStreamTrack. (My expectation may not be reasonable.)

**Actual behavior:**

Calling enable() on a Local\*Track with an underlying MediaStreamTrack object with enabled already set to true appears to be a noop. Calling disable() on a Local\*Track with an underlying MediaStreamTrack object with enabled already set to false appears to be a noop.

### (3) Publishing disabled audio/video tracks

An audio track that is published while the underlying MediaStreamTrack is disabled is, on the remote side:
  - added and started
  - muted is false

A video track that is published while the underlying MediaStreamTrack is disabled is, on the remote side:
  - added but not started
  - muted is true

This seems related to (1) above. (And consistent with that behavior.) This causes problems in our code as described in (1) above.

**Software versions:**

 - [x] Browser(s): Chrome 67.0.3396.99
 - [x] Operating System: Mac OS X and Linux
 - [x] twilio-video.js: 1.11.1
 - [x] Third-party libraries (e.g., Angular, React, etc.): React/Redux

**Comment by markandrus** [comment]
Hi @kwindla,

Thanks for taking the time to write so clearly. Let me attempt to answer!

### (1) Calling disable() makes a video track unplayable

#### Interaction between `enabled` and `muted`

> Calling disable() on a LocalVideoTrack sets muted to true for the receiving side's native MediaStreamTrack

You're right. Moreover: whether or not the underlying remote MediaStreamTrack's `muted` property is set to `true` will depend on whether you are using a P2P or Group Room (and perhaps your browser's WebRTC implementation). [Here](https://w3c.github.io/webrtc-pc/#mediastreamtrack-network-use) is what the WebRTC spec says about `muted`:

> When one of the SSRCs for RTP source media streams received by an `RTCRtpReceiver` is removed either due to reception of a BYE or via timeout, it MUST queue a task to set the muted state of the corresponding `MediaStreamTrack` to `true`.

And [here](https://w3c.github.io/webrtc-pc/#rtcrtpsender-interface) is what WebRTC spec says about `enabled`:

> If `track` is ended, or if the track's output is disabled, i.e. the track is disabled and/or muted, the `RTCRtpSender` MUST send silence (audio), black frames (video) or a zero-information-content equivalent. In the case of video, the `RTCRtpSender` SHOULD send one black frame per second.

Taken together, according to WebRTC, just setting `enabled` to `false` on the underlying local MediaStreamTrack (which is exactly what LocalVideoTrack's `disable` API does, plus some extra signaling) should not change the remote MediaStreamTrack's `muted` property. This is because black frames will still be sent for that SSRC. In fact, if you check `muted` in a P2P Room with two Chrome Participants, you'll see `muted` remains `false`. This scenario is basically the same as the "Chrome-to-Chrome PeerConnection" case you describe.

However, if you run the same test in a Group Room, you'll see `muted` changes to `true`. That's because Twilio's media server stops transmitting black frames for muted SSRCs. It cuts them off. We do this to reduce bandwidth and to work around [this bug](https://bugs.chromium.org/p/webrtc/issues/detail?id=5727), which causes Chrome to use too much bandwidth when publishing a disabled simulcast track.

#### Interaction between `muted` and `play()`

> **Actual behavior:**
>
> Calling disable() on a LocalVideoTrack results in any stream containing the track not being playable on the remote side. The promise returned by `play()` neither resolves or rejects. The muted field of the MediaStreamTrack on the receiving side is set to true. A video that is already playing a stream containing this track is able to continue playing (audio tracks remain audible). But new calls to `play()` do not complete until `enable()` is called on the sending side.

I've reviewed [the WHATWG spec for `play()`](https://html.spec.whatwg.org/multipage/media.html#dom-media-play), and it says this:

> 7. Otherwise, if the media element's `readyState` attribute has the value `HAVE_FUTURE_DATA` or `HAVE_ENOUGH_DATA`, take pending play promises and queue a task to resolve pending play promises with the result.

I expect that, due to the behavior described above, if you have a MediaStream containing a single `muted` MediaStreamTrack, `readyState` will not equal `HAVE_FUTURE_DATA` or `HAVE_ENOUGH_DATA`, and so `play()` will not resolve. Similarly, if you have an empty MediaStream, `play()` does not resolve:

```js
(async () => {
  const audio = document.createElement('audio');
  audio.srcObject = new MediaStream();
  await audio.play();
  console.log('Success');  // Never prints  
})().catch(console.error)
```

But should `play()` resolve if the MediaStream contains at least one non-`muted` MediaStreamTrack? Perhaps there is a browser bug or `play()` is under-specced.

Nevertheless, let me ask this: does your use case require that you call `play()`? I've looked at Twilio's various "dog food" apps for testing twilio-video.js, and none of them use `play()`. Instead, we always set `autoplay` to `true` and re-assign the `srcObject` (we actually don't use AudioTrack's and VideoTrack's `attach` API, since it's tricky to integrate with React). Can you do something similar?

#### Potential resolution/workaround

Use `autoplay` instead of `play()` and ignore the underlying remote MediaStreamTrack's `enabled` and `muted` properties?

### (2) Local\*Track enable()/disable() behavior depend on the enabled state of the underlying MediaStreamTrack

> **Actual behavior:**
>
> Calling enable() on a Local\*Track with an underlying MediaStreamTrack object with enabled already set to true appears to be a noop. Calling disable() on a Local\*Track with an underlying MediaStreamTrack object with enabled already set to false appears to be a noop.

This is expected, based on the algorithm `enable()` and `disable()` use. It's basically

* Get new value of `enabled` from the API call. Call it `nextEnabled`.
* Get the current value of `enabled` from the MediaStreamTrack. Call it `prevEnabled`.
* If `nextEnabled !== prevEnabled`, signal the change in `isEnabled`.

> So there are parts of our code that call enable/disable directly on MediaStreamTrack objects that the twilio-video library also holds a reference to.

Ah, so the algorithm above won't work if you change the underlying `enabled` state outside of Local\*Track, since `nextEnabled` and `prevEnabled` will match already. We could perhaps augment Local\*Track with a new property—something like `lastEnabled`. Any reads to the underlying MediaStreamTrack's `enabled` property should update `lastEnabled` and, any time Local\*Track observes a drift, it should signal the change in `isEnabled`. I'm not sure if it's a good idea, though. I will propose this to the team, though. If you have any ideas, feel free to share.

### (3) Publishing disabled audio/video tracks

I think this is because Twilio's media server does not perform the same optimization for audio tracks as it does with video tracks. I will confirm with the media server team.

**Comment by markandrus** [comment]
> I think this is because Twilio's media server does not perform the same optimization for audio tracks as it does with video tracks. I will confirm with the media server team.

My teammates on the Media Server Team tell me they apply the same "switch-off" for disabled AudioTracks as VideoTracks, so I'm not sure why Chrome acts differently. It's worth digging in to.

Can you please tell me if the workaround I proposed works for you?

**Comment by kwindla** [comment]
Mark, thank you for the detailed reply. It was very helpful. I also really appreciate the pointer to the simulcast track-disabled bitrate bug. I'd missed seeing that, and it explains some things I've been trying to figure out while testing simulcast.

The workaround you suggested -- using `autoplay` instead of calling `play()` -- is a good idea. That isn't a quick fix for us, though, because we bundle audio and video tracks together into streams as soon as we have references to both tracks, and play media using a single `<video>` element per meeting participant. (Quite a bit of our code predates `addTrack()` being available in Chrome, and is overdue for a rewrite.)

I've done a partial refactor of our code to separate out audio and video tracks, and to use separate `<audio>` and `<video>` elements for incoming media playback. This seems to work around the track-disabled issue, and is probably what we should be doing anyway, in 2018! A bit of spelunking in the js console suggests that Hangouts Meet uses this approach to play peer streams, so that's a good sign.

We will likely still call `play()` directly, instead of configuring media elements to `autoplay`. We have some code that handles playback being blocked by the new Chrome autoplay policy which is well tested enough (and tricky enough to test) that we're reluctant to change it.

Again, thanks for the reply.
